src/Glimpse.Common/Broker/DefaultMessageConverter.cs

```diff
@@ -18,36 +33,88 @@ public DefaultMessageConverter(JsonSerializer jsonSerializer, IContextData<Messa
         }
 
         public IMessage ConvertMessage(object payload)
-        { 
-            var message = new Message();
-            message.Id = Guid.NewGuid();
-            message.Type = payload.GetType().FullName;
-            message.Payload = _jsonSerializer.Serialize(payload);
-            message.Context = _contextData.Value;
+        {
+            var message = new Message
+            {
+                Id = Guid.NewGuid(),
+                Types = GetTypes(payload),
+                Payload = _jsonSerializer.Serialize(payload),
+                Context = _contextData.Value,
+                Indices = GetIndices(payload)
+            };
 
-            ProcessIndices(payload, message);
-            ProcessTags(payload, message);
+            message.Payload = _jsonSerializer.Serialize(message);
 
             return message;
-        } 
+        }
 
-        private void ProcessIndices(object payload, Message message)
+        private static IEnumerable<string> GetTypes(object payload)
         {
-            var indicesPayload = payload as IMessageIndices;
-            if (indicesPayload?.Indices != null)
+            var typeInfo = payload.GetType().GetTypeInfo();
+
+            return typeInfo.BaseTypes(includeSelf: true)
+                .Concat(typeInfo.ImplementedInterfaces)
+                .Except(_exclusions)
+                .Select(t => t.KebabCase());
+        }
+
+        private static IReadOnlyDictionary<string, object> GetIndices(object payload)
```

Comment:
Makes sense, but wondering if it is worth pulling the Type stuff into a provider. I could see that being modified or wrapped by people. Imagine that if for a given message type (i.e. BeginMessage) they wanted to add an extra type to the list, a provider here would be a low level way of doing that. They could let be `base` call happen and then add to the list that is returned.